server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java

```diff
@@ -99,61 +100,56 @@ private void extensionsDiscovery() throws IOException {
         if (!FileSystemUtils.isAccessibleDirectory(extensionsPath, logger)) {
             return;
         }
-        for (final Path plugin : PluginsService.findPluginDirs(extensionsPath)) {
-            try {
-                PluginInfo pluginInfo = PluginInfo.readFromProperties(plugin);
-
-                // TODO: Add unit tests for commented section below
+        List<Path> pluginDirs = PluginsService.findPluginDirs(extensionsPath);
 
-                /**
+        Set<Extension> extensions = new HashSet<Extension>();
+        if (!pluginDirs.isEmpty()) {
+            try {
+                extensions = readFromExtensionsYml(extensionsPath.resolve("extensions.yml")).getExtensions();
+            } catch (Exception e) {
+                throw new IllegalStateException("Could not read from extensions.yml");
+            }
+        }
+        for (final Path plugin : pluginDirs) {
+            if (Files.isDirectory(plugin)) {
                 try {
-                    List<Extension> extensions = readFromExtensionsYml(extensionsPath.toString()).getExtensions();
-                    for(Extension extension : extensions) {
-                        extensionsSet.add(
-                            new DiscoveryExtension(
-                                extension.getName(),
-                                "id",
-                                extension.getEphemeralId(),
-                                extension.getHostName(),
-                                extension.getHostAddress(),
-                                new TransportAddress(TransportAddress.META_ADDRESS, Integer.parseInt(extension.getPort())),
-                                null,
-                                Version.fromString(extension.getVersion()),
-                                pluginInfo
-                    )
-                );
+                    PluginInfo pluginInfo = PluginInfo.readFromProperties(plugin);
+                    String pluginName = pluginInfo.getName();
+                    for (Extension extension : extensions) {
+                        if (extension.getName().equals(pluginName)) {
+                            extensionsSet.add(
```

Comment:
Would be nice to add a debug log statement after this indicating which extension was loaded.